core/src/main/java/io/grpc/Metadata.java

```diff
@@ -449,7 +449,8 @@ private static String validateName(String n) {
 
     private Key(String name) {
       this.originalName = checkNotNull(name);
-      this.name = validateName(this.originalName.toLowerCase(Locale.ROOT));
+      // Intern the result for faster string identity checking.
+      this.name = validateName(this.originalName.toLowerCase(Locale.ROOT)).intern();
```

Comment:
No, Key.equals calls name.equals.  This is the String class equals which checks pointers before characters:

```
    public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        ...
```